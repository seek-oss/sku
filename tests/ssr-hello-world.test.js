const path = require('path');
const fs = require('fs/promises');

const {
  runSkuScriptInDir,
  waitForUrls,
  getAppSnapshot,
  startAssetServer,
} = require('@sku-private/test-utils');
const gracefulSpawn = require('../lib/gracefulSpawn');

const skuBuildConfig = require('@sku-fixtures/ssr-hello-world/sku-build.config.js');
const skuStartConfig = require('@sku-fixtures/ssr-hello-world/sku-start.config.js');

const appDir = path.dirname(
  require.resolve('@sku-fixtures/ssr-hello-world/sku-build.config.js'),
);

const getTestConfig = (skuConfig) => ({
  backendUrl: `http://localhost:${skuConfig.serverPort}`,
  targetDirectory: path.join(appDir, skuConfig.target),
});

describe('ssr-hello-world', () => {
  describe('start', () => {
    const { backendUrl } = getTestConfig(skuStartConfig);
    let server;

    beforeAll(async () => {
      server = await runSkuScriptInDir('start-ssr', appDir, [
        '--config=sku-start.config.js',
      ]);
      await waitForUrls(backendUrl);
    });

    afterAll(async () => {
      await server.kill();
    });

    it('should start a development server', async () => {
      const snapshot = await getAppSnapshot(backendUrl);
      expect(snapshot).toMatchSnapshot();
    });
  });

  describe('build', () => {
    const { backendUrl, targetDirectory } = getTestConfig(skuBuildConfig);
    let closeAssetServer;

    beforeAll(async () => {
      await runSkuScriptInDir('build-ssr', appDir, [
        '--config=sku-build.config.js',
      ]);

      closeAssetServer = await startAssetServer(4000, targetDirectory);
    });

    afterAll(() => {
      closeAssetServer();
    });

    describe('default port', () => {
      let server;

      beforeAll(async () => {
        server = gracefulSpawn('node', ['server'], {
          cwd: targetDirectory,
          stdio: 'inherit',
        });
        await waitForUrls(backendUrl);
      });

      afterAll(async () => {
        await server.kill();
      });

      it('should generate a production server based on config', async () => {
        const snapshot = await getAppSnapshot(backendUrl);
        expect(snapshot).toMatchSnapshot();
      });

      it("should invoke the provided 'onStart' callback", async () => {
        const pathToFile = path.join(targetDirectory, 'started.txt');
        const startedFile = await fs.readFile(pathToFile, {
          encoding: 'utf-8',
        });

        expect(startedFile).toMatchInlineSnapshot(
          `"Server started, here's your callback"`,
        );
      });
    });

    describe('custom port', () => {
      const customPort = 7654;
      const customPortUrl = `http://localhost:${customPort}`;
      let server;

      beforeAll(async () => {
        server = gracefulSpawn('node', ['server', '--port', customPort], {
          cwd: targetDirectory,
          stdio: 'inherit',
        });
        await waitForUrls(customPortUrl);
      });

      afterAll(async () => {
        await server.kill();
      });

      it('should generate a production server running on custom port', async () => {
        const snapshot = await getAppSnapshot(customPortUrl);
        expect(snapshot).toMatchSnapshot();
      });
    });
  });
});
