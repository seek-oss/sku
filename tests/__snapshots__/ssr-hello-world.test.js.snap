// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ssr-hello-world build custom port should generate a production server running on custom port 1`] = `
SCRIPTS: [
  "http://localhost:4000/runtime-21ba7ee762f9d6616967.js",
  "http://localhost:4000/497-dc030950b95c64882531.js",
  "http://localhost:4000/main-2bee8904f80654a0d455.js",
  "https://code.jquery.com/jquery-3.5.0.slim.min.js",
]
CSS: [
  "http://localhost:4000/main-cb8848ae92e1e9d08128.css",
]
SOURCE HTML: <!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <meta
      http-equiv="Content-Security-Policy"
      content="script-src 'self' http://localhost:4000 https://some-cdn.com https://code.jquery.com 'sha256-WHrb12igXgDsoQpp4ijb4PgZBl4AXNijY5y96KkF2hY=';"
    >
    <link
      data-chunk="main"
      rel="stylesheet"
      href="styles[0]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[0]"
      crossorigin="anonymous"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[1]"
      crossorigin="anonymous"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[2]"
      crossorigin="anonymous"
    >
  </head>
  <body>
    <div id="app">
      <div class="eRNyDeb">
        <div class="MYunKdE">
          Hello World
        </div>
        <img src="http://localhost:4000/a85d2645cc4d80a20b5b.png">
      </div>
    </div>
    <script src="scripts[3]">
    </script>
    <script>
      console.log('Hi');
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
      crossorigin="anonymous"
    >
      []
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
      crossorigin="anonymous"
    >
      {"namedChunks":[]}
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[0]"
      crossorigin="anonymous"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[1]"
      crossorigin="anonymous"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[2]"
      crossorigin="anonymous"
    >
    </script>
  </body>
</html>
POST HYDRATE DIFFS: NO DIFF
`;

exports[`ssr-hello-world build default port should generate a production server based on config 1`] = `
SCRIPTS: [
  "http://localhost:4000/runtime-21ba7ee762f9d6616967.js",
  "http://localhost:4000/497-dc030950b95c64882531.js",
  "http://localhost:4000/main-2bee8904f80654a0d455.js",
  "https://code.jquery.com/jquery-3.5.0.slim.min.js",
]
CSS: [
  "http://localhost:4000/main-cb8848ae92e1e9d08128.css",
]
SOURCE HTML: <!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <meta
      http-equiv="Content-Security-Policy"
      content="script-src 'self' http://localhost:4000 https://some-cdn.com https://code.jquery.com 'sha256-WHrb12igXgDsoQpp4ijb4PgZBl4AXNijY5y96KkF2hY=';"
    >
    <link
      data-chunk="main"
      rel="stylesheet"
      href="styles[0]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[0]"
      crossorigin="anonymous"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[1]"
      crossorigin="anonymous"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[2]"
      crossorigin="anonymous"
    >
  </head>
  <body>
    <div id="app">
      <div class="eRNyDeb">
        <div class="MYunKdE">
          Hello World
        </div>
        <img src="http://localhost:4000/a85d2645cc4d80a20b5b.png">
      </div>
    </div>
    <script src="scripts[3]">
    </script>
    <script>
      console.log('Hi');
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
      crossorigin="anonymous"
    >
      []
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
      crossorigin="anonymous"
    >
      {"namedChunks":[]}
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[0]"
      crossorigin="anonymous"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[1]"
      crossorigin="anonymous"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[2]"
      crossorigin="anonymous"
    >
    </script>
  </body>
</html>
POST HYDRATE DIFFS: NO DIFF
`;

exports[`ssr-hello-world start should start a development server 1`] = `
SCRIPTS: [
  "http://127.0.0.1:8100/runtime.js",
  "http://127.0.0.1:8100/vendors-node_modules_pnpm_loadable_component_5_15_3_react_17_0_2_node_modules_loadable_compon-7a1f2b.js",
  "http://127.0.0.1:8100/main.js",
  "https://code.jquery.com/jquery-3.5.0.slim.min.js",
]
CSS: [
  "http://127.0.0.1:8100/main.css",
]
SOURCE HTML: <!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <meta
      http-equiv="Content-Security-Policy"
      content="script-src 'self' http://127.0.0.1:8100 https://some-cdn.com https://code.jquery.com 'sha256-WHrb12igXgDsoQpp4ijb4PgZBl4AXNijY5y96KkF2hY=' 'unsafe-eval';"
    >
    <link
      data-chunk="main"
      rel="stylesheet"
      href="styles[0]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[0]"
      crossorigin="anonymous"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[1]"
      crossorigin="anonymous"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[2]"
      crossorigin="anonymous"
    >
  </head>
  <body>
    <div id="app">
      <div class="lessStyles__root___eRNyDeb">
        <div class="lessStyles__nested___MYunKdE">
          Hello World
        </div>
        <img src="http://127.0.0.1:8100/a85d2645cc4d80a20b5b.png">
      </div>
    </div>
    <script src="scripts[3]">
    </script>
    <script>
      console.log('Hi');
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
      crossorigin="anonymous"
    >
      []
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
      crossorigin="anonymous"
    >
      {"namedChunks":[]}
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[0]"
      crossorigin="anonymous"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[1]"
      crossorigin="anonymous"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[2]"
      crossorigin="anonymous"
    >
    </script>
  </body>
</html>
POST HYDRATE DIFFS: NO DIFF
`;
